feat: add test location metadata option

MIJOTHY opened the pull request:
Certain tools support making use of file/line/col metadata from junit
xml files. This commit adds an option for adding this metadata to the
junit output.

This is perhaps somewhat arbitrary, compounded by there not being an
official junit schema, but I'd like to throw an initial version out there to see 
if you'd be willing to support it.

<!--

Thank you for your contribution! Please also think about (where applicable)

- the CHANGELOG, you can add an entry at the top underneath "Unreleased"
- the README and other documentation
- the tests, at least run them yourself before submitting (usually a `bin/kaocha` will do)

-->

MIJOTHY commented on src/kaocha/plugin/junit_xml.clj:
```diff
@@ -86,21 +86,38 @@
+  [test-file]
+  (let [abs-path #(.getAbsolutePath %)]
+    (-> test-file
+        (io/resource (deref Compiler/LOADER))
```
I don't like directly referencing `Compiler/LOADER` here, as it's depending on an implementation detail IMO. However, omitting it didn't work as expected (`(io/resource test-file) => nil`).  

This works (brittly) because in `kaocha` we can see:
* [`add-classpath`](https://github.com/lambdaisland/kaocha/blob/a505b88ede250135990934e3d1d3b3cf32fede88/src/kaocha/classpath.clj#L37-L51), which is used for adding tests to the classpath, is only ever called with its [single-arity form](https://github.com/lambdaisland/kaocha/search?q=add-classpath)
* The single-arity form of `add-classpath` uses the nullary form of [`classloader-hierarchy`](https://github.com/lambdaisland/kaocha/blob/a505b88ede250135990934e3d1d3b3cf32fede88/src/kaocha/classpath.clj#L18-L24), which in turn binds [`Compiler/LOADER`](https://github.com/lambdaisland/kaocha/blob/a505b88ede250135990934e3d1d3b3cf32fede88/src/kaocha/classpath.clj#L11-L16) to a `DynamicClassLoader` when it's not already bound.
* This will then have the test paths [added](https://github.com/lambdaisland/kaocha/blob/a505b88ede250135990934e3d1d3b3cf32fede88/src/kaocha/classpath.clj#L49), along with any other loaders that have addable classpaths

The situations in which this breaks are:
1. `Compiler/LOADER` is already bound to a classloader that doesn't satisfy [`modifiable-classloader?`](https://github.com/lambdaisland/kaocha/blob/a505b88ede250135990934e3d1d3b3cf32fede88/src/kaocha/classpath.clj#L30), so it never has its classpath added to. I don't know when this would typically happen.
1. `kaocha` updates its internals, which breaks this plugin. Again, I don't know how likely this is, but at least this would be caught by tests when version bumping.

The improvement options I can think of are:
1. Use the same logic as is used for the [classloader determination](https://github.com/lambdaisland/kaocha/blob/a505b88ede250135990934e3d1d3b3cf32fede88/src/kaocha/classpath.clj#L46-L47) to get the loader for this function, perhaps submitting a PR to kaocha to expose this logic. To be fair, it looks like we can just use any classloader that satisfies [`modifiable-classloader?`](https://github.com/lambdaisland/kaocha/blob/a505b88ede250135990934e3d1d3b3cf32fede88/src/kaocha/classpath.clj#L30), to avoid hard-coding a particular one. On the other hand, because of the steps I outlined
2. Do something else completely that I'm missing

Any thoughts welcome here.

plexus commented:
> Certain tools

Which tools are we talking about? Like you said there's no standardized junit.xml schema, so I think it would be worthwhile to document which tools we are targeting, as well as starting to keep track of the differences between tools and documenting that somewhere. The existing code has largely been written with CircleCI in mind.

> Compiler/LOADER

You're right that relying on this is not ideal. The add-classpath code in Kaocha is the result of much trial and error on how to reliably add paths to the classpath so Clojure picks them up. I've learned a lot about classloaders in the meantime, which resulted in [lambdaisland/classpath](https://github.com/lambdaisland/classpath) and three blog posts 

- https://lambdaisland.com/blog/2021-08-25-classpath-is-a-lie
- https://lambdaisland.com/blog/2021-11-03-making-nrepl-cider-more-dynamic-1
- https://lambdaisland.com/blog/2021-11-24-making-nrepl-cider-more-dynamic-2

I think the right thing to do here is to change Kaocha core to also set/unset the context classloader, that way a regular `io/resource` call should work as expected. Note that this is only an issue when people don't have their `:test-paths` already on the classpath. Many people will also include them in their `deps.edn` under a `:test` alias.

I don't think we need to block this on that though, we're unlikely to stop binding `Compiler/LOADER`.

MIJOTHY commented:
Thanks for the links re. classloaders.

Some tools that produce/consume the location metadata include:
* [action-junit-report](https://github.com/mikepenz/action-junit-report): uses this metadata to produce annotations on GitHub files (GitHub Actions CI). This is my main interest.
![image](https://user-images.githubusercontent.com/10106320/165512119-53c8c872-8af9-4c54-865b-4ddf0d1fa4fe.png)
* [pytest](https://docs.pytest.org/en/latest/how-to/output.html?highlight=junit#creating-junitxml-format-files): pytest can produce junit xml files, which have `file` and `line` properties. This is apparently part of the `xunit1` format, but I can't find other places where this is described.

MIJOTHY commented:
Hey @plexus, any further thoughts on this? If it's something you'd be willing to support, I can tidy up the PR a bit. Namely:
* Allowing for `:file` to be `nil`, instead of crashing when it's not there
* Adding something to the README
* Updating the changelog and other admin stuff

plexus commented:
Yes, that would be great! Seems we need to upgrade the CircleCI orbs, @alysbrooks should be able to give some pointers for that. A section in the README listing these different tools would be great, then we have a place where we can document caveats and differences in the future as well.

MIJOTHY commented:
>     * Allowing for `:file` to be `nil`, instead of crashing when it's not there

I've backtracked on this for now; do you know of any expected situations where `file` would be `nil`? I'm thinking it may be better to crash loudly than to silently omit this information, which could be hard to debug. 
 
>     * Adding something to the README
> 
>     * Updating the changelog and other admin stuff

Have added these now, and consolidated some bits in a `CI Integration` section.

MIJOTHY commented on src/kaocha/plugin/junit_xml.clj:
```diff
@@ -151,18 +173,25 @@
                 "Save the test results to a Ant JUnit XML file."
 
                 "--junit-xml-omit-system-out"
-                "Do not add captured output to junit.xml"]))
+                "Do not add captured output to junit.xml"
+
+                "--junit-xml-add-location-metadata"
+                "Add line, column, and file attributes to tests in junit.xml"]))
```
Interestingly, when running `bin/kaocha --help`, I don't see these available, and they are considered errors by the CLI:
<img width="814" alt="image" src="https://user-images.githubusercontent.com/10106320/166892133-5bb27216-0073-4745-91d4-c07c17e8c5aa.png">

MIJOTHY commented:
> Yes, that would be great! Seems we need to upgrade the CircleCI orbs, @alysbrooks should be able to give some pointers for that. A section in the README listing these different tools would be great, then we have a place where we can document caveats and differences in the future as well.

@alysbrooks any pointers for sorting this out?

plexus commented:
Can you give me push access to your branch? I'll see if I can sort it out.

MIJOTHY commented:
> Can you give me push access to your branch? I'll see if I can sort it out.

Invited as a collaborator, congratulations.

MIJOTHY commented:
> Can you give me push access to your branch? I'll see if I can sort it out.

Any luck with this?

MIJOTHY commented:
Is there still interest in merging this functionality? I'm keen to get rid of my fork and depend on the 'real' upstream.

plexus commented:
Sorry to keep you hanging, @MIJOTHY. We've all been extraordinarily busy, and there were a few yaks to shave to get this build to green again. I've fixed the build on main and rebased. This should be good to go in now.

MIJOTHY commented:
> Sorry to keep you hanging, @MIJOTHY. We've all been extraordinarily busy, and there were a few yaks to shave to get this build to green again. I've fixed the build on main and rebased. This should be good to go in now.

No worries, much appreciated for being open to the change and the feedback!
